### PR TITLE
[Compression v2] Auto-Compress Pruner

### PR DESCRIPTION
**Pull request by J-shang**
~Need discussion about the interface about `admm_params` in `AutoCompressPruner.__init__()`, ref to the usage in AutoCompressPruner example, it has too many params to pass in... And ADMM will have more complex interfaces in the future...~

TODO:
- [x] test & doc

Known Issue:
Loss increases during multi-GPU training.

### CODE REVIEW

File: docs/en_US/Compression/v2_pruning_algo.rst [hunk 1]
```diff
@@ -397,3 +398,43 @@ User configuration for Simulated Annealing Pruner
 **PyTorch**
 
 .. autoclass:: nni.algorithms.compression.v2.pytorch.pruning.SimulatedAnnealingPruner
+
+Auto Compress Pruner
+--------------------
+
+For each round, AutoCompressPruner prune the model for the same sparsity to achive the overall sparsity:
```
Comment (QuanluZhang) [inline docs/en_US/Compression/v2_pruning_algo.rst:405]: "for the same ..." -> for a fixed sparsity ratio (e.g., 20%) to achieve the overall sparsity (e.g., 80%).

File: docs/en_US/Compression/v2_pruning_algo.rst [hunk 2]
```diff
@@ -397,3 +398,43 @@ User configuration for Simulated Annealing Pruner
...
+
+.. code-block:: bash
+
+       1. Generate sparsities distribution using SimulatedAnnealingPruner
+       2. Perform ADMM-based structured pruning to generate pruning result for the next round.
```
Comment (QuanluZhang) [inline docs/en_US/Compression/v2_pruning_algo.rst:410]: why "for the next round"?

File: docs/en_US/Compression/v2_pruning_algo.rst [hunk 3]
```diff
@@ -397,3 +398,43 @@ User configuration for Simulated Annealing Pruner
...
+        'iterations': 10,
+        'training_epochs': 1
+    }
+    sa_params = {
+        'evaluator': evaluator
+    }
+    pruner = AutoCompressPruner(model, config_list, 10, admm_params, sa_params, finetuner=finetuner)
+    pruner.compress()
+    _, model, masks, _, _ = pruner.get_best_result()
+
```
Comment (QuanluZhang) [inline docs/en_US/Compression/v2_pruning_algo.rst:434]: would be better to link an example

File: nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+class AutoCompressPruner(IterativePruner):
+    """
+    Parameters
+    ----------
+    model : Module
+        The origin unwrapped pytorch model to be pruned.
+    config_list : List[Dict]
+        The origin config list provided by the user. Note that this config_list is directly config the origin model.
+        This means the sparsity provided by the origin_masks should also be recorded in the origin_config_list.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py:49]: what do you want to express with sentences?

File: nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    total_iteration : int
+        The total iteration number.
+    evaluator : Callable[[Module], float]
+        Evaluate the pruned model and give a score.
+    admm_params : Dict
+        The parameters pass to the ADMMPruner.
+
+        - trainer : Callable[[Module, Optimizer, Callable].
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py:57]: why not simply provide a link to ADMMPruner?

File: nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            Supported pruning algorithm ['level', 'l1', 'l2', 'fpgm', 'slim', 'apoz', 'mean_activation', 'taylorfo', 'admm'].
+        - pruning_params : Dict. Default: `{}`.
+            If the pruner corresponding to the chosen pruning_algorithm has extra parameters, put them as a dict to pass in.
+
+    log_dir : str
+        The log directory use to saving the result, you can find the best result under this folder.
+    keep_intermediate_result : bool
+        If keeping the intermediate result, including intermediate model and masks during each iteration.
+    finetuner : Optional[Callable[[Module], None]]
+        The finetuner handled all finetune logic, use a pytorch module as input, will be called in each iteration.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py:93]: handled -> handles, use -> take

Comment (QuanluZhang) [reply nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py:93]: ", will be ..." -> ". It will be called at the end of each iteration, usually for neutralizing the accuracy loss brought by the pruning in this iteration."

File: nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    speed_up : bool
+        If set True, speed up the model in each iteration.
+    dummy_input : Optional[torch.Tensor]
+        If `speed_up` is True, `dummy_input` is required for trace the model in speed up.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py:97]: "for trace" -> "for tracing"

File: nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+            If the pruner corresponding to the chosen pruning_algorithm has extra parameters, put them as a dict to pass in.
+
+    log_dir : str
+        The log directory use to saving the result, you can find the best result under this folder.
+    keep_intermediate_result : bool
+        If keeping the intermediate result, including intermediate model and masks during each iteration.
+    finetuner : Optional[Callable[[Module], None]]
+        The finetuner handled all finetune logic, use a pytorch module as input, will be called in each iteration.
+    speed_up : bool
+        If set True, speed up the model in each iteration.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py:95]: -> If set True, speed up the model at the end of each iteration to make the pruned model compact.

File: nni/algorithms/compression/v2/pytorch/pruning/iterative_pruner.py [hunk 9]
```diff
@@ -227,6 +224,11 @@ class SimulatedAnnealingPruner(IterativePruner):
         Cool down rate of the temperature.
     perturbation_magnitude : float
         Initial perturbation magnitude to the sparsities. The magnitude decreases with current temperature.
+    pruning_algorithm : str
+        Supported pruning algorithm ['level', 'l1', 'l2', 'fpgm', 'slim', 'apoz', 'mean_activation', 'taylorfo', 'admm'].
+        This iterative pruner will use the chosen corresponding pruner to prune the model in each iteration.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/iterative_pruner.py:229]: chosen corresponding pruner -> chosen one-shot pruner

File: nni/algorithms/compression/v2/pytorch/pruning/iterative_pruner.py [hunk 10]
```diff
@@ -227,6 +224,11 @@ class SimulatedAnnealingPruner(IterativePruner):
...
+    pruning_params : Dict
+        If the pruner corresponding to the chosen pruning_algorithm has extra parameters, put them as a dict to pass in.
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/iterative_pruner.py:231]: "the pruner corresponding to the chosen pruning_algorithm" -> "the chosen one-shot pruner"

File: docs/en_US/Compression/v2_pruning_algo.rst (hunk 1 above)
Comment (J-shang) [reply docs/en_US/Compression/v2_pruning_algo.rst]: change to

For total iteration number :math:`N`, AutoCompressPruner prune the model that survive the previous iteration for a fixed sparsity ratio (e.g., :math:`1-{(1-0.8)}^{(1/N)}`) to achieve the overall sparsity (e.g., :math:`0.8`)

File: docs/en_US/Compression/v2_pruning_algo.rst (hunk 2 above)
Comment (J-shang) [reply docs/en_US/Compression/v2_pruning_algo.rst]: change to

for the next iteration

File: docs/en_US/Compression/v2_pruning_algo.rst (hunk 3 above)
Comment (J-shang) [reply docs/en_US/Compression/v2_pruning_algo.rst:434]: add it

File: nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py (hunk 4 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py:49]: if the model already has `0.2` sparsity by `origin_masks` and the user wants a total `0.8` sparsity, set `sparsity=0.8`, not `0.6`

File: nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py (hunk 5 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py:57]: because not all `ADMMPruner` parameters are expected to be set.

File: nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py (hunk 6 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py]: fix it

File: nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py (hunk 7 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py]: fix it

File: nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py (hunk 8 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py]: fix it

File: nni/algorithms/compression/v2/pytorch/pruning/iterative_pruner.py (hunk 9 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/iterative_pruner.py:232]: we don't explain what is one-shot pruner in v2

File: nni/algorithms/compression/v2/pytorch/pruning/iterative_pruner.py (hunk 10 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/iterative_pruner.py:234]: same as previous

**Review by Fiascolsy — APPROVED** [review]

**Review by liuzhe-lz — APPROVED** [review]

File: nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+                         origin_config_list=origin_config_list,
+                         origin_masks=origin_masks,
+                         log_dir=log_dir,
+                         keep_intermediate_result=keep_intermediate_result)
+
+    def _iterative_pruner_reset(self, model: Module, config_list: List[Dict] = [], masks: Dict[str, Dict[str, Tensor]] = {}):
+        self.iterative_pruner.task_generator._log_dir = Path(self._log_dir_root, 'SA')
+        self.iterative_pruner.reset(model, config_list=config_list, masks=masks)
+
+    def allocate_sparsity(self, new_config_list: List[Dict], model: Module, masks: Dict[str, Dict[str, Tensor]]):
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py:34]: this function already tried the returned `config_list`, right?

File: nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py (hunk 4 above)
Comment (QuanluZhang) [reply nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py:49]: recorded -> counted

Comment (QuanluZhang) [reply nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py:49]: how could user provide `origin_masks`?

File: nni/algorithms/compression/v2/pytorch/pruning/iterative_pruner.py (hunk 10 above)
Comment (QuanluZhang) [reply nni/algorithms/compression/v2/pytorch/pruning/iterative_pruner.py:234]: -> If the chosen pruning_algorithm has extra parameters,

File: nni/algorithms/compression/v2/pytorch/pruning/tools/task_generator.py [hunk 12]
```diff
@@ -43,13 +43,15 @@ def __init__(self, total_iteration: int, origin_model: Module, origin_config_lis
...
         """
-        self.current_iteration = 0
-        self.target_sparsity = config_list_canonical(origin_model, origin_config_list)
         self.total_iteration = total_iteration
-
-        super().__init__(origin_model, origin_config_list=self.target_sparsity, origin_masks=origin_masks,
+        super().__init__(origin_model, origin_config_list=origin_config_list, origin_masks=origin_masks,
                          log_dir=log_dir, keep_intermediate_result=keep_intermediate_result)
 
+    def reset(self, model: Module, config_list: List[Dict] = [], masks: Dict[str, Dict[str, Tensor]] = {}):
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/task_generator.py:50]: should we call `reset` in init?

File: nni/algorithms/compression/v2/pytorch/pruning/tools/task_generator.py [hunk 13]
```diff
@@ -97,6 +100,9 @@ def generate_tasks(self, task_result: TaskResult) -> List[Task]:
     def generate_config_list(self, target_sparsity: List[Dict], iteration: int, compact2origin_sparsity: List[Dict]) -> List[Dict]:
         raise NotImplementedError()
 
+    def allocate_sparsity(self, new_config_list: List[Dict], model: Module, masks: Dict[str, Dict[str, Tensor]]):
+        return new_config_list
```
Comment (QuanluZhang) [inline nni/algorithms/compression/v2/pytorch/pruning/tools/task_generator.py:104]: why we need this function? directly return an input argument?

File: nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py (hunk 11 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py:34]: Yes, SA Pruner already tried this `config_list` and found this `config_list` has the highest score from `evaluator`.

File: nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py [hunk 14]
```diff
@@ -720,12 +720,15 @@ def trainer(model: Module, optimizer: Optimizer, criterion: Callable[[Tensor, Te
     def __init__(self, model: Module, config_list: List[Dict], trainer: Callable[[Module, Optimizer, Callable], None],
                  optimizer: Optimizer, criterion: Callable[[Tensor, Tensor], Tensor], iterations: int, training_epochs: int):
         self.trainer = trainer
+        # TODO: handle optimizer here will case additional memory use, need improve, also in WeightTrainerBasedDataCollector
```
[the remark this answers is not in the document]
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py:723]: The optimizer will reference the parameters in the old model, and these old parameters may not be released when the `model` in pruner is changed.

File: nni/algorithms/compression/v2/pytorch/pruning/tools/task_generator.py (hunk 12 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/task_generator.py:50]: yes, `reset` is called in `super().__init__(...)`

File: nni/algorithms/compression/v2/pytorch/pruning/tools/task_generator.py (hunk 13 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/tools/task_generator.py:104]: Yes, this change is for AutoCompressTaskGenerator, it can apply `SimulatedAnnealingPruner.compress()` in this function.

File: nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py (hunk 4 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/auto_compress_pruner.py:49]: sorry, I forgot I have removed this API in iterative pruner for simplifying, users need to use the scheduler to support `origin_masks` right now.
I will remove this line.

File: nni/algorithms/compression/v2/pytorch/pruning/iterative_pruner.py (hunk 10 above)
Comment (J-shang) [reply nni/algorithms/compression/v2/pytorch/pruning/iterative_pruner.py]: fix

**Review by QuanluZhang — APPROVED** [review]
